src/test/java/wolox/training/controllers/BookControllerTest.java

```diff
@@ -0,0 +1,94 @@
+package wolox.training.controllers;
+
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
+import static org.mockito.ArgumentMatchers.any;
+
+import java.util.Arrays;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mockito;
+import org.springframework.boot.test.autoconfigure.jdbc.AutoConfigureTestDatabase;
+import org.springframework.boot.test.autoconfigure.jdbc.AutoConfigureTestDatabase.Replace;
+import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
+import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest;
+import org.springframework.boot.test.mock.mockito.MockBean;
+import org.springframework.test.context.junit4.SpringRunner;
+import wolox.training.models.Book;
+import wolox.training.repositories.BookRepository;
+
+
+@RunWith(SpringRunner.class)
+@AutoConfigureMockMvc
+@WebMvcTest(BookController.class)
+@AutoConfigureTestDatabase(replace = Replace.NONE)
+public class BookControllerTest extends BaseControllerTest {
+
+  @MockBean
+  private BookRepository mockBookRepository;
+  private Book book = new Book();
+  private String bookAsJson = "{\"id\":1,\"author\":\"Author\",\"image\":\"La imagen\",\"title\":\"El título\",\"subtitle\":\"El subtítulo\",\"publisher\":\"El publisher\",\"year\":\"1989\",\"pages\":250,\"isbn\":\"ISBN\",\"users\":[]}";
+
+  @Before
+  public void setUp() {
+    book.setAuthor("Author");
+    book.setImage("La imagen");
+    book.setIsbn("ISBN");
+    book.setPages(250);
+    book.setPublisher("El publisher");
+    book.setSubtitle("El subtítulo");
+    book.setTitle("El título");
+    book.setYear("1989");
+    book.setId(1L);
+  }
+
+  @Test
+  public void whenFindById_thenReturnBook() throws Exception {
+    Mockito.when(mockBookRepository.findById(book.getId()))
+        .thenReturn(java.util.Optional.ofNullable(book));
+    getAndExpect("/api/books/" + book.getId(), status().isOk(), bookAsJson);
+  }
+
+  @Test
+  public void whenFindAll_thenReturnBooks() throws Exception {
+    Mockito.when(mockBookRepository.findAll()).thenReturn(Arrays.asList(book));
+    getAndExpect("/api/books", status().isOk(), "[" + bookAsJson + "]");
```

Comment:
This is why I made the BaseControllerTest thing, now all those asserts are one-liners. Is there a better way to do this? Something I'm not seeing?

Reply:
Usually we copy and paste those lines but it's true that is duplicated code :P I like what you did with the BaseController, niiice! :)
I think you can define a new class with the `@Configuration` annotation and make it extend from [WebMvcConfigurer](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/web/servlet/config/annotation/WebMvcConfigurer.html), so there you can configure some defaults values. But this goes beyond this training :P I think the way you did it is great!

Reply:
I'll check it out anyway, it could be interesting